shell/browser/api/electron_api_view.cc

```diff
@@ -335,13 +339,46 @@ void View::SetBackgroundColor(std::optional<WrappedSkColor> color) {
   view_->SetBackground(color ? views::CreateSolidBackground(*color) : nullptr);
 }
 
+void View::SetBorderRadius(int radius) {
+  border_radius_ = radius;
+  ApplyBorderRadius();
+}
+
+void View::ApplyBorderRadius() {
+  if (!border_radius_.has_value() || !view_)
+    return;
+
+  auto size = view_->bounds().size();
+
+  // Restrict border radius to the constraints set in the path builder class.
+  // If the constraints are exceeded, the builder will crash.
+  int radius;
+  {
+    float r = border_radius_.value() * 1.f;
+    r = std::min(r, size.width() / 2.f);
+    r = std::min(r, size.height() / 2.f);
+    r = std::max(r, 0.f);
+    radius = std::floor(r);
+  }
+
+  // RoundedRectCutoutPathBuilder has a minimum size of 32 x 32.
+  if (radius > 0 && size.width() >= 32 && size.height() >= 32) {
+    auto builder = ash::RoundedRectCutoutPathBuilder(gfx::SizeF(size));
+    builder.CornerRadius(radius);
+    view_->SetClipPath(builder.Build());
+  } else {
+    view_->SetClipPath(SkPath());
```

Comment:
Any chance we could expose SetClipPath more directly? I'd generally like to err on the side of power over ease of use in the Views API.

Reply:
The difficulty here is that the native views use a different code path for faster clipping
https://source.chromium.org/chromium/chromium/src/+/main:ui/views/controls/native/native_view_host_aura.cc;l=349;drc=8b6aaad8027390ce6b32d82d57328e93f34bb8e5

Arguably the WebContentsView border radius is the dominant use case. The clipping for the underlying View allows us to also clip the background color.

Reply:
Hm, afaict this code here doesn't use the fast path, right?

Reply:
`WebContentsView` is using [SetCornerRadii](https://github.com/electron/electron/blob/7cd05180309467469a2f608de4c833a2551047a5/shell/browser/api/electron_api_web_contents_view.cc#L83-L84) which uses the fast path. This is required to apply a mask to the underlying `views::NativeViewHost`.

This results in two implementations of `setBorderRadius`
- `View`: uses `SetClipPath` to mask rendering
- `WebContentsView`: uses `SetCornerRadii` to apply the fast path render clipping to the WebView
  - Also calls into the View's mask rendering so its background color gets a mask applied as well.

Reply:
If we feel strongly about exposing a more powerful `SetClipPath` API, we could always introduce one in the future and reimplement `View.prototype.setBorderRadius` using it in an internal JS API.

Are we comfortable unblocking this? @nornagon @electron/wg-api

Reply:
Fair enough! I'll unblock this.